src/thor/trace_route_action.cc

```diff
@@ -394,92 +395,105 @@ thor_worker_t::map_match(Api& request) {
+            discontinuity_edges_iter->second == path_edge_itr->edgeid) {
+          // start a new one
+          edge_groups.emplace_back(edge_group_t{path_edge_itr, path_edge_itr, {}, {}});
           ++discontinuity_edges_iter;
         }
-        m_temp_disjoint_edge_groups.back().first.emplace_back(back_edge_info);
+        // remember where the last one ended
+        std::get<1>(edge_groups.back()) = path_edge_itr;
       }
 
-      // here we mark edges in the path that get "upgraded" to break points
-      // because they occur just before a discontinuity
-      auto discontinuity_point_iter = match_results.cbegin();
-      for (auto& disjoint_edge_group : m_temp_disjoint_edge_groups) {
-        const auto& last_edge_in_group = disjoint_edge_group.first.back();
-        // in each edge group, we only care about the last edge where the discontinuity happens
-        while (discontinuity_point_iter->edgeid != last_edge_in_group.edgeid) {
-          ++discontinuity_point_iter;
-        }
-        // for the last edge, the last match result on this edge is where discontinuity happens
-        while (discontinuity_point_iter->edgeid == last_edge_in_group.edgeid) {
-          ++discontinuity_point_iter;
+      // here we mark the original shape points that just after and just before the
+      // discontinuities so that in the path they get "upgraded" to break points
+      auto match_result_itr = match_results.cbegin();
+      for (auto& edge_group : edge_groups) {
+        // we know the discontinuity happens on this edge so we just need to
+        // find the first match result on this edge
+        auto first_edge_id = std::get<0>(edge_group)->edgeid;
+        while (match_result_itr != match_results.cend() &&
+               match_result_itr->edgeid != first_edge_id) {
+          ++match_result_itr;
         }
+        std::get<2>(edge_group) = match_result_itr;
 
-        disjoint_edge_group.second = discontinuity_point_iter - 1;
+        // we know the discontinuity happens on this edge so we just need to
+        // find the last match result on this edge
+        auto last_edge_id = std::get<1>(edge_group)->edgeid;
+        while (match_result_itr != match_results.cend() && match_result_itr->edgeid != last_edge_id) {
+          ++match_result_itr;
+        }
+        while (match_result_itr != match_results.cend() && match_result_itr->edgeid == last_edge_id) {
+          ++match_result_itr;
+        }
+        std::get<3>(edge_group) = std::prev(match_result_itr);
       }
 
       // The following logic put break points (matches results) on edge candidates to form legs
       // logic assumes the both match results and edge candidates are topologically sorted in correct
       // order
-      auto leg_origin_iter = match_results.cbegin();
-      for (const auto& disjoint_edge_group : m_temp_disjoint_edge_groups) {
+      auto origin_match_result = match_results.cbegin();
+      for (const auto& edge_group : edge_groups) {
```

Comment:
made a bunch of name changes to stick with the original nomenclature of the code and slightly reduce the lenght of variables. multiline autoformatting is brutal in terms of human parsibility imho